### DISCUSSION TITLE
esp32s3 - combining tinyuf2 and application

### DISCUSSION THREAD

**Discussion by rtborg** [adafruit/tinyuf2#328] [Q&A] [answered]
Is it possible to merge tinyuf2 and application binaries in a single binary, which can then be flashed via DFU or UART?

**Comment by tannewt** [comment, marked as answer]
Yes, you can use objcopy to merge binaries.

**Reply by rtborg** [reply]
I tried using `esptool.py merge_bin` to combine all binaries. I placed the application at address 0x10000 (I'm using 8MB flash size), flashed and the device goes straight into tinyUF2.  That failed, so I tried placing the application at address 0x210000.  

After that also failed, I flashed tinyUF2 and then uploaded the application separately. Then I read the contents of the flash and compared with the manually merged file. With the exception of the NVS data, which is present when I dump the flash, files appear the same.   

Do you have any thoughts on why is that happening?

**Reply by tannewt** [reply]
@makermelissa May know. I think she's done it for the web loader.

**Reply by makermelissa** [reply]
I'm not sure what you're trying to do. For the web loader, it either flashed tinyuf2 (using combined.bin file from the release) and then installed the circuitpython UF2 into flash memory or it just installed the circuitpython.bin file, but it was never combined with tinyuf2.

**Reply by rtborg** [reply]
I'm trying to use `idf.py write-flash` for flashing the tinyuf2 binaries and the application binary in one step. It fails - the device stays in tinyUF2.   
The flash dump of a working tinyUF2 + application reveals some data at 0xe000 (ota_data_initial.bin). I don't have access to my PC now, but tomorrow I'll check if tinyUF2 is reading/writing there.

**Reply by hathach** [reply]
you will need to mark the ota0 partition as valid for it to boot. It is simply 0x0001 ircc. If you use Ardduino it is called boot_app0 bin

**Reply by rtborg** [reply]
That did the job, thank you for the input.  

I examined the data at 0xe000, and in addition to 0x01 there are some bytes at 0xE01C, namely 0x9a984347. These bytes are found in both boot_app0.bin, and in the flash dump of a working tinyuf2+application. I initially assumed that was CRC of the ota0 partition, but that's likely wrong.   

In any case, the bootloader binaries can be combined with application binary, and then flashed together.
